core/templates/dev/head/forms/formBuilder.js

```diff
@@ -747,9 +726,383 @@ oppia.config(['$provide', function($provide) {
+        var blockOverlayRule = ' div(oppia-rte-component-overlay);';
+        // Put all the rules together.
+        var extraAllowedContentRules = componentRule +
+                                       inlineWrapperRule +
+                                       blockWrapperRule +
+                                       blockOverlayRule;
+
+        // The button corresponding to a widget has the same name,
+        // with a capitalized first letter.
+        var buttonNames = [];
+        pluginNames.forEach(function(pluginName) {
+          buttonNames.push(pluginName.charAt(0).toUpperCase() +
+                           pluginName.slice(1, pluginName.length));
+          // '-' adds the vertical divider between toolbar buttons.
+          buttonNames.push('-');
+        });
+        // Remove the last '-' from the list.
+        buttonNames.pop();
+
+        // Initialize ckeditor.
+        var ck = CKEDITOR.inline(el[0].children[0].children[1], {
+          extraPlugins: 'widget,lineutils,sharedspace,' + pluginNames.join(','),
+          startupFocus: true,
+          extraAllowedContent: extraAllowedContentRules,
+          sharedSpaces: {
+            // Select the directive template's empty div.
+            top: el[0].children[0].children[0]
+          },
+          skin: 'bootstrapck,/third_party/static/ckeditor-bootstrapck/',
+          toolbar: [
+            {
+              name: 'history',
+              items: ['Undo', '-', 'Redo']
+            },
+            {
+              name: 'basicstyles',
+              items: ['Bold', '-', 'Italic', '-', 'RemoveFormat']
+            },
+            {
+              name: 'paragraph',
+              items: [
+                'NumberedList', '-',
+                'BulletedList', '-',
+                'Outdent', '-',
+                'Indent'
+              ]
+            },
+            {
+              name: 'rtecomponents',
+              items: buttonNames
+            }
+          ]
+        });
+
+        /**
+         * A RegExp for matching rich text components. It has 3 capture groups:
+         * p1: The entire opening tag, including brackets and all attributes.
+         * p2: Just the name of the tag, e.g. "oppia-noninteractive-math"
+         * p3: Just the name of the component, e.g. "math"
+         */
+        var componentRe = (
+          /(<(oppia-noninteractive-(.+?))\b[^>]*>)[\s\S]*?<\/\2>/g
+        );
+
+        /**
+         * Before data is loaded into CKEditor, we need to wrap every rte
+         * component in a span (inline) or div (block).
+         * For block elements, we add an overlay div as well.
+         */
+        var wrapComponents = function(html) {
+          if (html === undefined) {
+            return html;
+          }
+          return html.replace(componentRe, function(match, p1, p2, p3) {
+            if (rteHelperService.isInlineComponent(p3)) {
+              return '<span type="oppia-noninteractive-' + p3 + '">' +
```

Comment:
This looks similar to the componentTemplate stuff on lines 758 and 762. Should they be consolidated in order to avoid skew?